### PR TITLE
feat: Support entity type mappings

### PR DESCRIPTION
**Pull request by jimmycallin**
- [ ] I have added automatic tests where applicable
- [ ] The PR title is suitable as a release note
- [ ] The PR contains a description of what has been changed
- [ ] The description contains manual test instructions

## Changes

Suggestion for alternative implementation of type mapping that don't couple ts-schema-generator and ftrack-api.

Should be as simple as:

```
import TypeMapping from "./__generated__/schema.ts"
import { Session } from @ftrack/api

const session = new Session<TypeMapping>(...);

const user = await session.query<"User">("...");

```

This allows easier extension of the types in case they are wrong or the user want to do something weird

### CODE REVIEW

File: source/project_schema.ts [hunk 1]
```diff
@@ -18,8 +18,8 @@ import type { Data, QueryResponse } from "./types.js";
  *
  * @memberof project_schema
  */
-export function getStatuses(
-  session: Session,
+export function getStatuses<TEntityTypeMap extends Record<string, any>>(
+  session: Session<TEntityTypeMap>,
```
Comment (jimmycallin) [inline source/project_schema.ts:22]: this should automatically get the entitytypemap inferred from the provided session, make sure it works so you don't have to specify it

File: source/operation.ts [hunk 2]
```diff
@@ -109,13 +112,13 @@ export function query(expression: string): QueryOperation {
  * @param  {string} expression API query expression
  * @return {Object}            API operation
  */
-export function search({
+export function search<TEntityType>({
```
Comment (jimmycallin) [inline source/operation.ts:115]: TEntityType is automatically inferred from the options type, so no need to specify it explicitly (please verify)

Comment (gismya) [reply source/operation.ts:115]: That's clever if it works. Needs to be tested. The only issue I see is a lack of verification of the TEntitityType in the Options.

File: source/session.ts [hunk 3]
```diff
@@ -48,25 +48,29 @@ const ENCODE_DATETIME_FORMAT = "YYYY-MM-DDTHH:mm:ss";
  * @class  Session
  *
  */
-export class Session {
+export class Session<
+  TEntityTypeMap extends Record<string, any> = DefaultEntityTypeMap,
```
Comment (gismya) [inline source/session.ts:52]: I have a feeling that the extends `Record<string,any>` will break all the type safety, setting everything to any even if we specify a TEntityTypeMap

Comment (jimmycallin) [reply source/session.ts:52]: you can look at the test/session.test.ts and toy with it a bit, seem to be working:
image

**Comment by ffMathy** [comment]
I love where this is going. Way better. Great work!

**Review by gismya — COMMENTED** [review]
Seems to be working well. Operations work slightly less well with this structure than the module augmentation one, but I prefer this way of doing it anyways. We just need to add some documentation on how to use it.

**Review by gismya — DISMISSED** [review]

**Comment by gismya** [comment]
Needs to be tested by integrating in our code base before releasing, but my initial tests all look good.

**Review by gismya — APPROVED** [review]

**Comment by ffMathy** [comment]
This makes me so happy! Thank you!

**Pull request merged**